packages/map/src/CanvasPins.ts

```diff
@@ -0,0 +1,351 @@
+    topLeftY_exists(): boolean;
+}
+
+CanvasPins.prototype.publish("allCircles", false, "boolean", "allCircles", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("showQuadtree", false, "boolean", "showQuadtree", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("useAveragePos", false, "boolean", "useAveragePos", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("shrinkFontToPin", true, "boolean", "shrinkFontToPin", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("enableClustering", true, "boolean", "enableClustering", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("useWeightedRadius", false, "boolean", "useWeightedRadius", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("radiusWeightMult", 0.5, "number", "radiusWeightMult", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("searchRectMult", 3, "number", "searchRectMult", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("bottomRightX", null, "number", "Bottom right x-value", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("bottomRightY", null, "number", "Bottom right y-value", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("topLeftX", null, "number", "Top left x-value", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("topLeftY", null, "number", "Top left y-value", null, { tags: ["Basic"], optional: true });
+
+CanvasPins.prototype.publish("pinHeight", 20, "number", "pinHeight", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("pinWidth", 20, "number", "pinWidth", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("pinFontFamily", "Arial", "string", "pinFontFamily", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("pinFontSize", 14, "number", "pinFontSize", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("arrowHeight", 8, "number", "arrowHeight", null, { tags: ["Basic"], optional: true });
+CanvasPins.prototype.publish("arrowWidth", 8, "number", "arrowWidth", null, { tags: ["Basic"], optional: true });
+
+export class Quadtree {
+    protected _pin_h = 0;
+    protected _pin_w = 0;
+    protected _visit_count: number = 0;
+    protected _tree: any;
+    constructor(extent: any, raw_data: any, pin_h: number, pin_w: number) {
+        this._pin_h = pin_h;
+        this._pin_w = pin_w;
+        this._tree = d3quadtree()
+            .extent(extent)
+            .addAll(raw_data);
+    }
+
+    searchRect(left: number, top: number, right: number, bottom: number): any {
+        const ret: any = [];
+        const pin_h = this._pin_h;
+        const pin_w = this._pin_w;
+        this._tree.visit((node: any, x1: number, y1: number, x2: number, y2: number) => {
+            this._visit_count++;
+            if (!node.length) {
```

Comment:
When will node be an array?  (or is my assumption that node is an object incorrect?)

Reply:
The `node` variable is a 4 element array when visiting "internal" nodes. The d3 quadtree documentation suggests using `!node.length` to distinguish between leaf and internal nodes.